buildtest/buildsystem/base.py

```diff
@@ -0,0 +1,558 @@
+           version of it. This is intended to log in the console for the user.
+
+           Parameters
+           ==========
+           key: The key to look up in the metadata
+           fmt: the format string to use
+        """
+        timestamp = self.metadata.get(key, "")
+        if timestamp:
+            timestamp = timestamp.strftime(fmt)
+        return timestamp
+
+    def show_prepare(self):
+        """Print basic run information to the user, if defined, before the run.
+        """
+        start_time = self.get_formatted_time("start_time")
+
+        print("{:_<80}".format(""))
+        print("{:>40} {}".format("start time:", self.metadata.get("start_time")))
+
+        if self.config_file:
+            print("{:>40} {}".format("configuration file:", self.config_file))
+        print("{:>40} {}".format("testdir:", self.metadata.get("testdir")))
+        print("{:>40} {}".format("testpath:", self.metadata.get("testpath")))
+        print("{:>40} {}".format("logpath:", self.metadata.get("logfile")))
+        print("{:_<80}".format(""))
+
+        print("\n\n")
+        print("{:<40} {}".format("STAGE", "VALUE"))
+        print("{:_<80}".format(""))
+
+    @run_wrapper
+    def dry_run(self):
+        """Akin to a build preview, we prepare and finish a run, but only
+           print the script to the screen without writing or running files.
+        """
+        # Dry run just prints the testing script
+        self._init_logger(to_file=False)
+        lines = self._get_test_lines()
+        print("\n".join(lines))
+
+    def _generate_build_id(self):
+        """Generate a build id based on the recipe name, type, and datetime
+        """
+        now = datetime.datetime.now().strftime("%m-%d-%Y-%H-%M-S")
+        return "%s_%s_%s" % (self.name, self.type, now)
+
+    def _write_test(self, lines=None):
+        """Given test metadata, write test content.
+        """
+        # If no lines provided, generate
+        if not lines:
+            lines = self._get_test_lines()
+
+        # '$HOME/.buildtest/testdir/<name>/<name>_<timestamp>.sh'
+        # This will put output (latest run) in same directory - do we want this?
+        testpath = os.path.expandvars(self.metadata["testpath"])
+        testdir = os.path.dirname(testpath)
+
+        self.logger.info(f"Opening Test File for Writing: {testpath}")
+
+        # Open the test file and write contents
+        with open(testpath, "w") as fd:
+            fd.write("\n".join(lines))
+
+        # Change permission of the file to executable
+        os.chmod(
+            testpath,
+            stat.S_IRWXU | stat.S_IRGRP | stat.S_IXGRP | stat.S_IROTH | stat.S_IXOTH,
+        )
+        print("{:<40} {}".format("[WRITING TEST]", "PASSED"))
+        return testpath
+
+    def _get_test_lines(self):
+        """Given test metadata, get test lines to write to file or show.
+        """
+        # Every test starts with cd to TESTDIR
+        lines = ["cd $TESTDIR"]
```

Comment:
I am now wondering if it makes sense to add ``cd $TESTDIR`` to all test-scripts. I think the purpose was to ensure any output, error or log files are not put in current directory which could be anywhere depending on where user ran the command. I think buildtest should correctly write all .out, .err, and .log and any other files in same or sub-directory where test exists.

Reply:
This would fundamentally change the behavior of the library, and basically eliminate the need to store test results in the buildtest home. I think this is a good idea, but since it's a bit of a refactor of the library, it should be addressed in another PR, with an issue opened first for discussion.